lib/Sema/CSDiag.cpp

```diff
@@ -8727,29 +8728,28 @@ bool FailureDiagnosis::diagnoseArchetypeAmbiguity() {
     // because type B would have no constraints associated with it.
     unsigned numConstraints = 0;
     {
-      llvm::SmallVector<Constraint *, 2> constraints;
+      llvm::SetVector<Constraint *> constraints;
       CS.getConstraintGraph().gatherConstraints(
-          tv, constraints, ConstraintGraph::GatheringKind::EquivalenceClass);
-
-      for (auto constraint : constraints) {
-        // We are not interested in ConformsTo constraints because
-        // such constraints specify restrictions on the archetypes themselves.
-        if (constraint->getKind() == ConstraintKind::ConformsTo)
-          continue;
+          tv, constraints, ConstraintGraph::GatheringKind::EquivalenceClass,
+          [&](Constraint *constraint) -> bool {
+            // We are not interested in ConformsTo constraints because
+            // such constraints specify restrictions on the archetypes
```

Comment:
I'd rather not say "archetypes" here; I think the reason we don't care about `conformsTo` constraints is that we can't derive any concrete type information from them.

Reply:
Sounds good! It's an existing comment I've moved here but I'll fix it up if perf test is green...